ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,529 @@
+
+	topTrieChecksum, err := storeTopLevelNodesAndTrieRoots(
+		tries, subTrieRootIndices, subTriesNodeCount, outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store top level tries: %w", err)
+	}
+
+	err = storeCheckpointSummary(subTrieChecksums, topTrieChecksum, outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store checkpoint summary: %w", err)
+	}
+
+	logger.Info().Msgf("checkpoint file has been successfully stored at %v/%v", outputDir, outputFile)
+
+	return nil
+}
+
+// 		1. version
+//		2. checksum of each part file (17 in total)
+// 		3. checksum of the main file itself
+func storeCheckpointSummary(
+	subTrieChecksums []uint32,
+	topTrieChecksum uint32,
+	outputDir string,
+	outputFile string,
+	logger *zerolog.Logger,
+) error {
+	closable, err := createWriterForCheckpointSummary(outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store checkpoint summary: %w", err)
+	}
+	defer func() {
+		closeErr := closable.Close()
+		// Return close error if there isn't any prior error to return.
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	// write version
+	version := encodeVersion(MagicBytes, VersionV6)
+	writer := NewCRC32Writer(closable)
+	_, err = writer.Write(version)
+	if err != nil {
+		return fmt.Errorf("cannot write checksum summary header: %w", err)
+	}
+
+	// TODO: write checksums
+
+	// write checksum to the end of the file
+	checksum := writer.Crc32()
+	_, err = writer.Write(encodeCRC32Sum(checksum))
+	if err != nil {
+		return fmt.Errorf("cannot write CRC32 checksum to checkpoint summary: %w", err)
+	}
+	return nil
+}
+
+func createWriterForCheckpointSummary(outputDir string, outputFile string, logger *zerolog.Logger) (io.WriteCloser, error) {
+	fullPath := filePathHeader(outputDir, outputFile)
+	if utilsio.FileExists(fullPath) {
+		return nil, fmt.Errorf("checkpoint file already exists at %v", fullPath)
+	}
+
+	return createClosableWriter(outputDir, logger, outputFile, fullPath)
+}
+
+// 17th part file contains:
+// 1. checkpoint version TODO
+// 2. checkpoint file part index TODO
+// 3. subtrieNodeCount TODO
+// 4. top level nodes
+// 5. trie roots
+// 6. node count
+// 7. trie count
+// 6. checksum
+func storeTopLevelNodesAndTrieRoots(
+	tries []*trie.MTrie,
+	subTrieRootIndices map[*node.Node]uint64,
+	subTriesNodeCount uint64,
+	outputDir string,
+	outputFile string,
+	logger *zerolog.Logger,
+) (uint32, error) {
+	// the remaining nodes and data will be stored into the same file
+	closable, err := createWriterForTopTries(outputDir, outputFile, logger)
+	if err != nil {
+		return 0, fmt.Errorf("could not create writer for top tries: %w", err)
+	}
+	defer func() {
+		closeErr := closable.Close()
+		// Return close error if there isn't any prior error to return.
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	writer := NewCRC32Writer(closable)
```

Comment:
Not sure how much crc32 is useful here: it is usually used for very small quantities of data.  Is there a downside of using a strong sha2 checksum here?

Reply:
We have been using crc32, I'm not sure how much difference they are, changing it would be out of scope of this PR.

But thanks for pointing this out. Worth to explore the difference.

Reply:
[Will be tracked and done by this issue](https://github.com/onflow/flow-go/issues/3302)